python/ray/data/impl/push_based_shuffle.py

```diff
@@ -0,0 +1,388 @@
+            block, meta = reduce_fn(*reduce_args, *mapper_outputs)
+            merged_outputs.append(block)
+        meta = BlockAccessor.for_block(block).get_metadata(
+            input_files=None, exec_stats=stats.build()
+        )
+        return [meta] + merged_outputs
+
+    @staticmethod
+    def _get_cluster_cpu_map():
+        nodes = ray.nodes()
+        # Map from per-node resource name to number of CPUs available on that
+        # node.
+        cpu_map = {}
+        for node in nodes:
+            resources = node["Resources"]
+            num_cpus = int(resources.get("CPU", 0))
+            if num_cpus == 0:
+                continue
+            cpu_map[node["NodeID"]] = num_cpus
+        return cpu_map
+
+    @staticmethod
+    def _compute_merge_task_options(
+        num_merge_tasks_per_round: int, merge_factor: int, cpu_map: Dict[str, int]
+    ):
+        """
+        Compute the task options needed to schedule merge tasks.
+
+        Each merge task should be grouped with `merge_factor` map tasks for
+        pipelining. These groups should then be spread across nodes according
+        to CPU availability for load-balancing. Across the rounds of merge
+        tasks, we also want to make sure that the i-th task from each round is
+        colocated on the same node to reduce data movement during the final
+        reduce stage.
+
+        Args:
+            num_merge_tasks_per_round: The total number of merge tasks in each
+                round.
+            merge_factor: The ratio of map : merge tasks.
+            cpu_map: A map from node ID to the number of CPUs available on that
+                node.
+        """
+        merge_to_node_map = []
+        merge_tasks_assigned = 0
+        node_ids = list(cpu_map)
+        leftover_cpu_map = {}
+        while merge_tasks_assigned < num_merge_tasks_per_round and cpu_map:
+            node_id = node_ids[merge_tasks_assigned % len(node_ids)]
+            # We group `merge_factor` map tasks with 1 merge task.
+            if cpu_map[node_id] >= merge_factor + 1:
+                cpu_map[node_id] -= merge_factor + 1
+                merge_to_node_map.append(node_id)
+                merge_tasks_assigned += 1
+            else:
+                leftover_cpu_map[node_id] = cpu_map.pop(node_id)
+                node_ids.remove(node_id)
+        while merge_tasks_assigned < num_merge_tasks_per_round:
+            node_id = max(leftover_cpu_map, key=leftover_cpu_map.get)
+            # We should have enough CPUs to schedule a round of merge tasks in
+            # parallel.
+            assert leftover_cpu_map[node_id] != 0
+            assert leftover_cpu_map[node_id] <= merge_factor
+            merge_to_node_map.append(node_id)
+            merge_tasks_assigned += 1
+            leftover_cpu_map[node_id] -= 1
+        merge_task_args = [
+            {"scheduling_strategy": NodeAffinitySchedulingStrategy(node_id, soft=True)}
```

Comment:
Should we also use NodeAffinitySchedulingStrategy to colocate `merge_factor` number of map tasks with the merge task?

Reply:
I think it is actually better to let Ray schedule the map tasks in case there are other factors like data locality being used to schedule those. I'll make a note of this in the code!